EIPS/eip-4361.md

```diff
@@ -0,0 +1,295 @@
+
+
+### Example message to be signed
+```
+service.org wants you to sign in with your Ethereum account:
+0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
+
+I accept the ServiceOrg Terms of Service: https://service.org/tos
+
+URI: https://service.org/login
+Version: 1
+Chain ID: 1
+Nonce: 328917
+Issued At: 2021-09-30T16:25:24Z
+Resources:
+- ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/
+- https://example.com/my-web2-claim.json
+```
+
+### Informal Message Template
+A Bash-like informal template of the full message is presented below for readability and ease of understanding. Field descriptions are provided in the following section. A full ABNF description is provided in the section thereafter.
+```
+${domain} wants you to sign in with your Ethereum account:
+${address}
+
+${statement}
+
+URI: ${uri}
+Version: ${version}
+Chain ID: ${chain-id}
+Nonce: ${nonce}
+Issued At: ${issued-at}
+Expiration Time: ${expiration-time}
+Not Before: ${not-before}
+Request ID: ${request-id}
+Resources:
+- ${resources[0]}
+- ${resources[1]}
+...
+- ${resources[n]}
+```
+### Message Field Descriptions
+- `domain` is the RFC 4501 `dnsauthority` that is requesting the signing.
+- `address` is the Ethereum address performing the signing conformant to capitalization encoded checksum specified in EIP-55 where applicable.
+- `statement` (optional) is a human-readable ASCII assertion that the user will sign, and it must not contain `'\n'`.
+- `uri` is an RFC 3986 URI referring to the resource that is the subject of the signing (as in the _subject of a claim_).
+- `version` is the current version of the `message`, which MUST be `1` for this specification.
+- `chain-id` is the EIP-155 Chain ID to which the session is bound, and the network where Contract Accounts must be resolved.
+- `nonce` is a randomized token used to prevent replay attacks, at least 8 alphanumeric characters.
+- `issued-at` is the ISO 8601 datetime string of the current time.
+- `expiration-time` (optional) is the ISO 8601 datetime string that, if present, indicates when the signed authentication message is no longer valid.
+- `not-before` (optional) is the ISO 8601 datetime string that, if present, indicates when the signed authentication message will become valid.
+- `request-id` (optional) is an system-specific identifier that may be used to uniquely refer to the sign-in request.
+- `resources` (optional) is a list of information or references to information the user wishes to have resolved as part of authentication by the relying party. They are expressed as RFC 3986 URIs separated by `"\n- "`.
+
+### ABNF Message Format
+
+The `message` to be signed MUST conform with the following Augmented Backus–Naur Form (ABNF, RFC 5234) expression (note that `%s` denotes case sensitivity for a string term, as per RFC 7405).
```

Comment:
Why is it preferable to have a new invented language and not some binary format, which is easier to parse (CBOR or even ERC-712)?

Is the only reason because the signed message will be "readily" displayed in wallets without implementing support for it? If so, I think hardware wallets may be a big exception to that where some other more easily parsable format is more likely to gain support.

Reply:
Hi @axic, thanks for your concern. That is indeed the main reason, so that applications may adopt this specification without full buy-in from wallet vendors or significant degradation of their user experiences today. In our proposal response to the RFP, we actually specified EIP-712 as the signing format, but found the user experience across many wallets to be much worse than using EIP-191. Formats like CBOR, EIP-712, protobuf, etc. indeed specify structuring for data, but hardware wallets have spotty support for presentation of EIP-712 requests to the users already (please see ongoing issues https://github.com/LedgerHQ/app-ethereum/issues/105 and https://github.com/trezor/trezor-firmware/pull/1568). We believe that if adoption of this specification relies on all wallet vendors upgrading how they do signing first, it is far less likely to see success. While pure technical merit is important to consider, adoption depends far more on talking to downstream users and understanding their concerns, what they are likely to truly adopt, etc. We have completed over 30 interviews towards this conclusion. We welcome anyone to let us know of any further user research in support or incongruent with this claim. Hope that helps!

Reply:
Hi, just wanted to follow up on this thread. I understand that people like structured signing and this decision may be unpopular, but we did the work and this is how it looks like in wallets today: https://blog.spruceid.com/sign-in-with-ethereum-wallet-support-eip-191-vs-eip-712/

Applications we spoke to are not happy to degrade their UX significantly, and people have little idea what they are signing if confronted with a JSON blob. Therefore, I think EIP-191 is still the right move here. 

If you'd like to consider CBOR, EIP-712, protobuf, etc., I would need some evidence that indeed broad wallet support will come for these soon so users can see what they're signing. We added a `version` term to allow this as an upgrade path in the future, but for this first iteration of the spec it seems premature. Thanks so much for your review and consideration.